src/Service/Bulk/BulkCreateUsersService.php

```diff
@@ -0,0 +1,294 @@
+    }
+
+    public function getOperationResultMessage(array $slugTotalUsers, array $userPrefix): string
+    {
+        $message = '';
+        $userPrefixString = implode(',', $userPrefix);
+
+        foreach ($slugTotalUsers as $slug => $totalUsers) {
+            $message .= sprintf(
+                "%s users created for line item %s for user prefix %s \n",
+                $totalUsers,
+                $slug,
+                $userPrefixString
+            );
+        }
+
+        return $message;
+    }
+
+
+    private function generateSlugData(array $lineItems): void
+    {
+        $lineItemSlugs = [];
+        foreach ($lineItems as $lineItem) {
+            $lineItemSlugs[$lineItem->getId()] = $lineItem->getSlug();
+        }
+        $this->lineItemSlugs = $lineItemSlugs;
+    }
+
+    private function createBulkUserAssignmentData(
+        array $userNameLastIndexes,
+        array $userGroupIds,
+        array $userPrefix,
+        int $userGroupAssignCount,
+        ?string $groupPrefix,
+        int $batchSize
+    ): array {
+        $slugWiseTotalUsersArray = [];
+        $automateCsvPath = $this->generatedUsersFilePath . date('Y-m-d');
+
+        if (!$this->filesystem->exists($automateCsvPath)) {
+            $this->filesystem->mkdir($automateCsvPath);
+        }
+
+        foreach ($userPrefix as $prefix) {
+            $csvPath = sprintf('%s/%s', $automateCsvPath, $prefix);
+            if (!$this->filesystem->exists($csvPath)) {
+                $this->filesystem->mkdir($csvPath);
+            }
+
+            foreach ($this->lineItemSlugs as $lineSlugs) {
+                $slugTotalUsersCreated = 0;
+                $csvFilename = sprintf('%s-%s.csv', $lineSlugs, $prefix);
+                $csvData = $assignmentCsvData = [];
+                foreach (range(1, $batchSize) as $batchIncrementValue) {
+                    $username = sprintf(
+                        '%s_%s_%d',
+                        $lineSlugs,
+                        $prefix,
+                        (
+                            (int)$userNameLastIndexes[$lineSlugs] + $batchIncrementValue
+                        )
+                    );
+                    $userPassword = $this->createUserPassword();
+                    $userGroupId = $groupPrefix
+                        ? $this->createUserGroupId($userGroupIds, $userGroupAssignCount)
+                        : '';
+
+                    $csvData[] = [$username, $userPassword,$userGroupId];
+                    $assignmentCsvData[] = [$username, $lineSlugs];
+
+                    $slugTotalUsersCreated++;
+                }
+
+                $this->csvWriter->writeCsvData(
+                    $lineSlugs,
+                    $prefix,
+                    $csvPath,
+                    $csvFilename,
+                    $automateCsvPath,
+                    $csvData,
+                    $assignmentCsvData
+                );
```

Comment:
Forgot during first review:
the problem with this piece of code and underlying method (`...->writeCsvData`) is in fact that it shouldn't be responsible for writing more than one CSV.
I would propose to change it and make more generic, and just simply call in there twice

Reply:
Small addition to my comment:
after thinking for a while, it seems that `BulkUserCreationCsvWriter` class is not needed at all - all the paths/data to be written by the underlying `CsvWriter` class can be prepared in this class.